[android] add an Android NDK Swift overlay

hyp opened the pull request:
This PR is based on the PR that adds the module map for the android NDK:

https://github.com/apple/swift/pull/72161

hyp committed:
[android] add a module map for Android NDK

This commit doesn't install them yet, they will be installed and a whole Android NDK module will be built in a follow-up commit

This module map covers the Bionic C standard library and other Posix headers used in the Android NDK

hyp committed:
[android] add an android NDK Swift overlay module, and use it instead of Glibc

compnerd commented on lib/ClangImporter/ClangIncludePaths.cpp:
```diff
@@ -186,7 +186,8 @@ static bool shouldInjectLibcModulemap(const llvm::Triple &triple) {
 static SmallVector<std::pair<std::string, std::string>, 2>
 getLibcFileMapping(ASTContext &ctx, StringRef modulemapFileName,
                    std::optional<StringRef> maybeHeaderFileName,
-                   const llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> &vfs) {
+                   const llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> &vfs,
+                   std::optional<StringRef> maybeSecondHeaderFileName = {}) {
```
Can we change the signature to instead be `std::optional<ArrayRef<StringRef>>` for the headers?

compnerd commented on lib/ClangImporter/ClangIncludePaths.cpp:
```diff
@@ -227,19 +228,23 @@ getLibcFileMapping(ASTContext &ctx, StringRef modulemapFileName,
-    Path actualHeaderPath = actualModuleMapPath;
-    llvm::sys::path::remove_filename(actualHeaderPath);
-    llvm::sys::path::append(actualHeaderPath, maybeHeaderFileName.value());
+  auto addHeader = [&](std::optional<StringRef> maybeHeaderFileName) {
```
What else can it be if it is not a header? The `maybe` here is about the fact that the file may not exist right?

compnerd commented on stdlib/public/Platform/SwiftAndroidNDK.h:
```diff
@@ -0,0 +1,92 @@
+#ifndef SWIFT_ANDROID_NDK_MODULE
```
Do we not include copyright headers?

compnerd commented on stdlib/public/Platform/SwiftBionic.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef SWIFT_BIONIC_MODULE
```
Likewise

hyp commented:
@swift-ci please test

hyp commented:
This needs some changes to build Foundation that @compnerd has as well.

hyp commented:
@swift-ci please test

hyp commented:
@swift-ci please test source compatibility

hyp committed:
android test fixes

finagolfin commented:
I [made some comments](https://github.com/apple/swift/commit/47863938e70ac807cbad984d99f6a22269cdbd14#r141745780) on your latest commit.

hyp committed:
add one more concurrency test fix

hyp committed:
update tests to use canImport(android) and make libc test for generalized

hyp commented:
@finagolfin addressed your review comments

finagolfin committed:
[android] Update tests to use new Android overlay and disable stdlib/Glibc.swift

finagolfin commented on stdlib/public/Differentiation/TgmathDerivatives.swift.gyb:
```diff
@@ -18,12 +18,14 @@ import Swift
   import WASILibc
 #elseif os(Windows)
   import CRT
+#elseif os(Android)
```
`canImport`

finagolfin commented:
Fixed some executable tests that were failing natively on Android and rebased, hyp#1.

finagolfin commented on stdlib/public/Platform/Platform.swift:
```diff
@@ -241,7 +241,7 @@ public var S_IFIFO: Int32 { return Int32(0x1000) }
 public var S_IWRITE: Int32 { return Int32(0x0080) }
 public var S_IEXEC: Int32  { return Int32(0x0040) }
-#else
+#elseif !os(Android)
```
Do we really need to do this? I understand that you are avoiding [this subsequent compilation error when trying to build against the resulting stdlib](https://github.com/finagolfin/swift-android-sdk/actions/runs/9242458261/job/25425332005#step:6:4072) and it does fix that:
```
swift-corelibs-foundation/Sources/Foundation/FileHandle.swift:324:39: error: ambiguous use of 'S_IFMT'
 322 | 
 323 |         let readBlockSize: Int
 324 |         if Int(statbuf.st_mode) & Int(S_IFMT) == Int(S_IFREG) {
     |                                       `- error: ambiguous use of 'S_IFMT'
 325 |             // TODO: Should files over a certain size always be mmap()'d?
 326 |             if options.contains(.alwaysMapped) {

Android.S_IFMT:1:12: note: found this candidate in module 'Android'
1 | public var S_IFMT: mode_t { get }
  |            `- note: found this candidate in module 'Android'

posix_filesystem.S_IFMT:1:12: note: found this candidate in module 'posix_filesystem'
1 | public var S_IFMT: Int32 { get }
  |            `- note: found this candidate in module 'posix_filesystem'
```
However, it then causes problems downstream when building projects like NIO, as these constants are imported as `Int32` from the C headers but `mode_t` is `UInt32` on Android AArch64:
```
swift-nio/Sources/NIOCore/FileHandle.swift:147:76: error: default argument value of type 'Int32' cannot be converted to type 'NIOPOSIXFileMode' (aka 'UInt32')
145 │         /// - parameters:
146 │         ///     - posixMode: `file mode` applied when file is created. Default permissions are: read and write for fileowner, read for owners group and others.
147 │         public static func allowFileCreation(posixMode: NIOPOSIXFileMode = defaultPermissions) -> Flags {
    │                                                                            ╰─ error: default argument value of type 'Int32' cannot be converted to type 'NIOPOSIXFileMode' (aka 'UInt32')
148 │             return Flags(posixMode: posixMode, posixFlags: O_CREAT)
149 │         }
```
It is convenient that these constants are typed `mode_t` on all non-Windows platforms: any way we can make this work? You would not need all the `mode_t` conversions on apple/swift-corelibs-foundation#4889 either if we got this working.

hyp replied:
Good question. I don't have a good idea for how to resolve this ambiguity, but perhaps we can special case these constants when they're imported on Android in the clang importer, and make them disfavored, so that Swift will always prefer the ones from the 'Android' module. what do you think of that?

finagolfin replied:
That seems extreme, must be a simpler way? I will try some things and let you know.

hyp replied:
Yeah, I think that will work! @finagolfin I will update the PR and swift-corelibs-foundation too

hyp replied:
Made some updates in https://github.com/apple/swift/pull/72161#issuecomment-2135610565 . @finagolfin do you mind rechecking NIO ?

finagolfin replied:
Will do.

finagolfin replied:
Hmm, I'm seeing native compilation errors in Termux after this change:
```
[220/223][ 98%][55.721s] Linking Swift shared library lib/libFoundation.so
FAILED: lib/libFoundation.so Sources/Foundation/CMakeFiles/Foundation.dir/AffineTransform.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Array.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/AttributedString/AttributedString.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/AttributedString/AttributedStringAttribute.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/AttributedString/AttributedStringCodable.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/AttributedString/AttributedStringRunCoalescing.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/AttributedString/AttributedString+Locking.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Boxing.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Bridging.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Bundle.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/ByteCountFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Calendar.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/CGFloat.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/CharacterSet.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Codable.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Collections+DataProtocol.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/ContiguousBytes.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/AttributedString/Conversion.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Data.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/DataProtocol.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Date.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/DateComponents.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/DateComponentsFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/DateFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/DateInterval.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/DateIntervalFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Decimal.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Dictionary.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/DispatchData+DataProtocol.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/EnergyFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/ExtraStringAPIs.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/FileHandle.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/FileManager.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/FileManager+POSIX.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/FileManager+Win32.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/FileManager+XDG.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Formatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/AttributedString/FoundationAttributes.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/FoundationErrors.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Host.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/IndexPath.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/IndexSet.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/ISO8601DateFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/JSONDecoder.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/JSONEncoder.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/JSONSerialization.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/JSONSerialization+Parser.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/LengthFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Locale.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/MassFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Measurement.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/MeasurementFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Morphology.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Notification.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NotificationQueue.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSArray.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSAttributedString.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCache.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCalendar.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCFArray.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCFBoolean.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCFCharacterSet.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCFDictionary.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCFSet.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCFString.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCFTypeShims.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCharacterSet.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCoder.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSComparisonPredicate.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSCompoundPredicate.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSConcreteValue.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSData+DataProtocol.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSData.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSDate.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSDateComponents.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSDecimalNumber.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSDictionary.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSEnumerator.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSError.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSExpression.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSGeometry.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSIndexPath.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSIndexSet.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSKeyedArchiver.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSKeyedArchiverHelpers.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSKeyedCoderOldStyleArray.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSKeyedUnarchiver.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSLocale.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSLock.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSLog.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSMeasurement.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSNotification.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSNull.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSNumber.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSObjCRuntime.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSObject.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSOrderedSet.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSPathUtilities.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSPersonNameComponents.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSPlatform.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSPredicate.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSRange.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSRegularExpression.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSSet.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSSortDescriptor.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSSpecialValue.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSString.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSStringAPI.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSSwiftRuntime.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSTextCheckingResult.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSTimeZone.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSURL.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSURLComponents.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSURLQueryItem.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSURLError.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSUUID.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NSValue.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/NumberFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Operation.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/PersonNameComponents.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/PersonNameComponentsFormatter.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Pointers+DataProtocol.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Port.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/PortMessage.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Process.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/ProcessInfo.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Progress.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/ProgressFraction.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/PropertyListEncoder.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/PropertyListSerialization.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/ReferenceConvertible.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/RunLoop.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Scanner.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/ScannerAPI.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Set.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Stream.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/String.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/StringEncodings.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Thread.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Timer.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/TimeZone.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/Unit.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/URL.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/URLComponents.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/URLQueryItem.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/URLResourceKey.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/UserDefaults.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/UUID.swift.o Sources/Foundation/CMakeFiles/Foundation.dir/WinSDK+Extensions.swift.o swift/Foundation.swiftmodule 
: && /data/data/com.termux/files/home/build/Ninja-Release/swift-android-aarch64/bin/swiftc -j 8 -num-threads 8 -emit-library -o lib/libFoundation.so -module-name Foundation -module-link-name Foundation -emit-module -emit-module-path swift/Foundation.swiftmodule -emit-dependencies -DDEPLOYMENT_RUNTIME_SWIFT -DFoundation_EXPORTS -DSWIFT_CORELIBS_FOUNDATION_HAS_THREADS -module-cache-path "/data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/module-cache" -O -wmo -Xcc -DSWIFT_CORELIBS_FOUNDATION_HAS_THREADS -Xfrontend -disable-autolink-framework -Xfrontend CoreFoundation -Xcc -F/data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64 -vfsoverlay /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64/dispatch-vfs-overlay.yaml -output-file-map Sources/Foundation/CMakeFiles/Foundation.dir/Release/output-file-map.json -I /data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers -I /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64/src/swift/swift -I /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64 -I /data/data/com.termux/files/home/swift-corelibs-libdispatch -I /data/data/com.termux/files/home/swift-corelibs-libdispatch/src -I /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64/src -I /data/data/com.termux/files/home/swift-corelibs-libdispatch/src/BlocksRuntime /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/AffineTransform.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Array.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/AttributedString/AttributedString.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/AttributedString/AttributedStringAttribute.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/AttributedString/AttributedStringCodable.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/AttributedString/AttributedStringRunCoalescing.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/AttributedString/AttributedString+Locking.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Boxing.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Bridging.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Bundle.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/ByteCountFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Calendar.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/CGFloat.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/CharacterSet.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Codable.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Collections+DataProtocol.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/ContiguousBytes.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/AttributedString/Conversion.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Data.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/DataProtocol.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Date.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/DateComponents.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/DateComponentsFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/DateFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/DateInterval.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/DateIntervalFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Decimal.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Dictionary.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/DispatchData+DataProtocol.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/EnergyFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/ExtraStringAPIs.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/FileHandle.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/FileManager.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/FileManager+POSIX.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/FileManager+Win32.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/FileManager+XDG.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Formatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/AttributedString/FoundationAttributes.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/FoundationErrors.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Host.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/IndexPath.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/IndexSet.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/ISO8601DateFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/JSONDecoder.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/JSONEncoder.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/JSONSerialization.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/JSONSerialization+Parser.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/LengthFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Locale.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/MassFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Measurement.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/MeasurementFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Morphology.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Notification.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NotificationQueue.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSArray.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSAttributedString.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCache.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCalendar.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCFArray.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCFBoolean.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCFCharacterSet.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCFDictionary.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCFSet.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCFString.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCFTypeShims.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCharacterSet.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCoder.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSComparisonPredicate.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSCompoundPredicate.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSConcreteValue.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSData+DataProtocol.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSData.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSDate.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSDateComponents.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSDecimalNumber.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSDictionary.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSEnumerator.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSError.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSExpression.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSGeometry.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSIndexPath.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSIndexSet.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSKeyedArchiver.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSKeyedArchiverHelpers.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSKeyedCoderOldStyleArray.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSKeyedUnarchiver.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSLocale.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSLock.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSLog.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSMeasurement.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSNotification.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSNull.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSNumber.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSObjCRuntime.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSObject.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSOrderedSet.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSPathUtilities.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSPersonNameComponents.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSPlatform.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSPredicate.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSRange.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSRegularExpression.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSSet.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSSortDescriptor.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSSpecialValue.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSString.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSStringAPI.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSSwiftRuntime.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSTextCheckingResult.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSTimeZone.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSURL.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSURLComponents.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSURLQueryItem.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSURLError.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSUUID.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NSValue.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/NumberFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Operation.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/PersonNameComponents.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/PersonNameComponentsFormatter.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Pointers+DataProtocol.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Port.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/PortMessage.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Process.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/ProcessInfo.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Progress.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/ProgressFraction.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/PropertyListEncoder.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/PropertyListSerialization.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/ReferenceConvertible.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/RunLoop.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Scanner.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/ScannerAPI.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Set.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Stream.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/String.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/StringEncodings.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Thread.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Timer.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/TimeZone.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Unit.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/URL.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/URLComponents.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/URLQueryItem.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/URLResourceKey.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/UserDefaults.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/UUID.swift /data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/WinSDK+Extensions.swift -no-toolchain-stdlib-rpath -Xlinker -soname -Xlinker libFoundation.so  -L /data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/lib  -L /data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/lib  -L /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64  -L /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64  -L /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64   -L /data/data/com.termux/files/usr/lib   -L /data/data/com.termux/files/usr/aarch64-linux-android/lib   -L /system/lib64 -Xlinker -rpath -Xlinker /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64:/data/data/com.termux/files/usr/lib:  lib/libCoreFoundation.a  lib/libuuid.a  /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64/libswiftDispatch.so  -llog  /data/data/com.termux/files/usr/lib/libicuuc.so  -ldl  /data/data/com.termux/files/usr/lib/libicui18n.so  /data/data/com.termux/files/usr/lib/libicudata.so  -lm  /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64/libdispatch.so  /data/data/com.termux/files/home/build/Ninja-Release/libdispatch-android-aarch64/libBlocksRuntime.so  -l:libunwind.a  -ldl  -lc  -l:libunwind.a  -ldl && :
<module-includes>:1:10: note: in file included from <module-includes>:1:
1 │ #include "Headers/CoreFoundation.h"
  │          ╰─ note: in file included from <module-includes>:1:
2 │ #include "Headers/CFPlugInCOM.h"
3 │ 

/data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/CoreFoundation.h:103:10: note: in file included from /data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/CoreFoundation.h:103:
101 │ #include <CoreFoundation/CFNotificationCenter.h>
102 │ 
103 │ #include <CoreFoundation/ForSwiftFoundationOnly.h>
    │          ╰─ note: in file included from /data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/CoreFoundation.h:103:
104 │ 
105 │ #endif /* ! __COREFOUNDATION_COREFOUNDATION__ */

/data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/ForSwiftFoundationOnly.h:607:18: error: definition of 'statx' must be imported from module 'posix_filesystem.linux_stat' before it is required
605 │ static inline int
606 │ _stat_with_btime(const char *filename, struct stat *buffer, struct timespec *btime) {
607 │     struct statx statx_buffer = {0};
    │                  ╰─ error: definition of 'statx' must be imported from module 'posix_filesystem.linux_stat' before it is required
608 │     *btime = (struct timespec) {0};
609 │ 

/data/data/com.termux/files/usr/include/linux/stat.h:59:8: note: definition here is not reachable
 57 │   __s32 __reserved;
 58 │ };
 59 │ struct statx {
    │        ╰─ note: definition here is not reachable
 60 │   __u32 stx_mask;
 61 │   __u32 stx_blksize;

<module-includes>:1:10: note: in file included from <module-includes>:1:
1 │ #include "Headers/CoreFoundation.h"
  │          ╰─ note: in file included from <module-includes>:1:
2 │ #include "Headers/CFPlugInCOM.h"
3 │ 

/data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/CoreFoundation.h:103:10: note: in file included from /data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/CoreFoundation.h:103:
101 │ #include <CoreFoundation/CFNotificationCenter.h>
102 │ 
103 │ #include <CoreFoundation/ForSwiftFoundationOnly.h>
    │          ╰─ note: in file included from /data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/CoreFoundation.h:103:
104 │ 
105 │ #endif /* ! __COREFOUNDATION_COREFOUNDATION__ */

/data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/ForSwiftFoundationOnly.h:610:87: error: use of undeclared identifier 'STATX_ALL'
608 │     *btime = (struct timespec) {0};
609 │ 
610 │     int ret = _statx(AT_FDCWD, filename, AT_SYMLINK_NOFOLLOW | AT_STATX_SYNC_AS_STAT, STATX_ALL, &statx_buffer);
    │                                                                                       ╰─ error: use of undeclared identifier 'STATX_ALL'
611 │     if (ret == 0) {
612 │         *buffer = (struct stat) {

<module-includes>:1:10: note: in file included from <module-includes>:1:
1 │ #include "Headers/CoreFoundation.h"
  │          ╰─ note: in file included from <module-includes>:1:
2 │ #include "Headers/CFPlugInCOM.h"
3 │ 

/data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/CoreFoundation.h:103:10: note: in file included from /data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/CoreFoundation.h:103:
101 │ #include <CoreFoundation/CFNotificationCenter.h>
102 │ 
103 │ #include <CoreFoundation/ForSwiftFoundationOnly.h>
    │          ╰─ note: in file included from /data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/CoreFoundation.h:103:
104 │ 
105 │ #endif /* ! __COREFOUNDATION_COREFOUNDATION__ */

/data/data/com.termux/files/home/build/Ninja-Release/foundation-android-aarch64/CoreFoundation.framework/Headers/ForSwiftFoundationOnly.h:628:37: error: use of undeclared identifier 'STATX_BTIME'
626 │         };
627 │         // Check that stx_btime was set in the response, not all filesystems support it.
628 │         if (statx_buffer.stx_mask & STATX_BTIME) {
    │                                     ╰─ error: use of undeclared identifier 'STATX_BTIME'
629 │             *btime = (struct timespec) {
630 │                 .tv_sec = statx_buffer.stx_btime.tv_sec,

/data/data/com.termux/files/home/swift-corelibs-foundation/Sources/Foundation/Bridging.swift:13:29: error: could not build C module 'CoreFoundation'
 11 │ //===----------------------------------------------------------------------===//
 12 │ 
 13 │ @_implementationOnly import CoreFoundation
    │                             ╰─ error: could not build C module 'CoreFoundation'
 14 │ 
 15 │ #if canImport(ObjectiveC)
ninja: build stopped: subcommand failed.
```
You don't see this when cross-compiling foundation?

hyp replied:
@finagolfin No, it builds when cross-compliing. However, the `linux/stat.h` include is only included whenever TARGET_ANDROID is set in the `ForSwiftFoundationOnly.h`:

```
#if TARGET_OS_ANDROID
#include <errno.h>
#include <sys/stat.h>
#include <sys/syscall.h>
#include <termios.h>
#include <linux/stat.h>
```

Is the `TARGET_OS_ANDROID` define set whenever we're building for termux?

finagolfin replied:
> However, the linux/stat.h include is only included whenever TARGET_ANDROID is set in the ForSwiftFoundationOnly.h

Where are you getting those includes from? I see [nothing like that in that header, which only pulls in that `linux/stat.h` header for Glibc](https://github.com/apple/swift-corelibs-foundation/blob/1dac1132d49b0ee4213ac89473019cae8a608b5c/CoreFoundation/Base.subproj/ForSwiftFoundationOnly.h#L86), which is not used on Android:
```
#if TARGET_OS_ANDROID
#include <errno.h>
#include <sys/stat.h>
#include <sys/syscall.h>
#include <termios.h>
#elif TARGET_OS_WASI
#include <fcntl.h>
#include <sys/stat.h>
// Define _WASI_EMULATED_MMAN here to use the emulated mman functions in
// Foundation-side without requiring transitive clients to define it.
#undef _WASI_EMULATED_MMAN
#define _WASI_EMULATED_MMAN
#include <sys/mman.h>
#elif TARGET_OS_LINUX
#include <errno.h>
#include <features.h>
#include <termios.h>

#ifdef __GLIBC_PREREQ
#if __GLIBC_PREREQ(2, 28) == 0
// required for statx() system call, glibc >=2.28 wraps the kernel function
#include <sys/syscall.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <linux/stat.h>
```
> Is the TARGET_OS_ANDROID define set whenever we're building for termux?

Yes, it should be, otherwise it would not build at all.

finagolfin replied:
Oh, I see, [you just modified your foundation pull to add it](https://github.com/apple/swift-corelibs-foundation/pull/4889/files#diff-472be7809a38fcf36eb7b8c41cb3c1a1afb235ef1cca057332ab1179ea685ab1R67). I was building against the previous version of that pull first, to make sure everything still worked, so I missed that.

I will update my foundation patch and try again.

hyp committed:
Fix canImport android

hyp committed:
Android: do not re-export macros from linux/stat.h from other modules

This lets us declare constants in the Android platform module using the expected mode_t type

hyp committed:
drop android stuff

hyp commented:
@ian-twilightcoder thanks for the feedback! I removed the let constants from Android.swift like you suggested,  and it works correctly now in terms of actually using these constants, although the deprecated warnings are now missing. but I have some follow-up comments/questions inline:
- For things like FLT_RADIX, the NDK's clang's builtin float.h isn't modularized yet , so we can't yet build _Builtin_float float.swift.gyb overlay for Swift. They're actually then imported from `_limits` module as the limits.h header in the NDK imports float.h from clang's builtins. Is it ok to just drop the overlay?
- For things like M_PI , could you elaborate on exactly do we need the _math module overlay? Is it just for deprecation warnings? Can it be just dropped?
- Also, how do you avoid ambiguity for something that's redeclared in Swift like M_PI if we have an overlay in Swift but we're also importing M_PI from the _math module?

hyp commented:
@swift-ci please test

hyp commented:
@swift-ci please test

finagolfin commented:
> I don't really know the history of deprecating the FLT_ and M_ constants and how important that is.

@stephentyrone, you added those warnings to the `Glibc` overlay: should we keep them in this new Android overlay, to nudge Swift devs to the Swift equivalents?

egorzhdan commented:
It might be easier to use API Notes to mark the constants as deprecated in Swift rather than making a new overlay module for it.

hyp commented:
@egorzhdan is it possible to use API notes for macros that are imported as constants? I thought that wasn't possible.

egorzhdan commented:
Oh right, I wasn't aware that those are macros. API notes are not going to work then.

hyp commented:
@swift-ci please test

hyp committed:
fix canImport clause in the MathConstants.swift testcase

hyp commented:
@swift-ci please test

hyp commented:
The deprecated warnings are tested in : 
- test/stdlib/FloatConstants.swift
- test/stdlib/MathConstants.swift

I'll check if I can get them working for Android. I'll see if I can create the right overlay module for them.

hyp commented:
@swift-ci please test macOS platform

hyp commented:
@swift-ci please test macOS platform

hyp commented:
@swift-ci please test macOS platform

finagolfin commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -273,12 +273,28 @@ add_custom_target(musl_modulemap DEPENDS ${musl_modulemap_target_list})
   if(NOT "${sdk}" STREQUAL "LINUX" AND
      NOT "${sdk}" STREQUAL "FREEBSD" AND
      NOT "${sdk}" STREQUAL "OPENBSD" AND
-     NOT "${sdk}" STREQUAL "ANDROID" AND
```
You'll also want to revert [my Android change below](https://github.com/apple/swift/commit/3f4109b3a26ad1cf51c7d66920ead67cb86bb7fc#diff-b1b3ce8fbe4a2f02397d6e5805ae0869fa7ada243daf0d58c97d7f73bccf0b7c).

finagolfin commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -376,6 +392,79 @@ add_custom_target(glibc_modulemap DEPENDS ${glibc_modulemap_target_list})
+      OUTPUT ${module_dir}/${android_modulemap_source} ${module_dir_static}/${android_modulemap_source}
+      COMMENT "Copying Android modulemap to resource directories")
+    add_custom_command_target(
+      copy_android_ndk_neader_resource
```
ndk_header

finagolfin commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -376,6 +392,79 @@ add_custom_target(glibc_modulemap DEPENDS ${glibc_modulemap_target_list})
+      OUTPUT ${module_dir}/${android_ndk_header_source} ${module_dir_static}/${android_ndk_header_source}
+      COMMENT "Copying Android NDK header to resource directories")
+    add_custom_command_target(
+      copy_android_bionic_neader_resource
```
bionic_header

finagolfin commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -376,6 +392,79 @@ add_custom_target(glibc_modulemap DEPENDS ${glibc_modulemap_target_list})
+                                 ${copy_android_bionic_neader_resource})
+    list(APPEND android_modulemap_target_list ${copy_android_modulemap_resource}
+                                              ${copy_android_ndk_neader_resource}
+                                              ${copy_android_bionic_neader_resource})
```
Correct these four also.

finagolfin reviewed (approved):
This looks good, once the deprecation warnings and other small nits I just raised are addressed.

hyp committed:
[stdlib] add a _math module overlay for Android for deprecated constants

hyp committed:
[cmake] fix android header name targets

hyp committed:
remove additional glibc android check from cmake

hyp commented:
@ian-twilightcoder I've added the '_math' module Swift overlay and the math constants are now properly deprecated from Swift . For float constants, it looks like using the '_Builtin_float' overlay works as well on Linux! Although at the moment on windows specifically I can't build it yet as the NDK headers doesn't ship an updated clang module map with '_Builtin_float', so it'll have to wait until an NDK update for the Windows-built Android SDK.

Does it look good to you?

hyp commented:
@swift-ci please test

hyp committed:
build CxxStdlib for Android too

hyp commented:
@swift-ci please test

hyp commented:
@swift-ci please test source compatibility

finagolfin commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -337,8 +367,7 @@ foreach(sdk ${SWIFT_SDKS})
     # If this SDK is a target for a non-native host, except if it's for Android
```
Please revert [my now incorrect comment also](https://github.com/apple/swift/commit/3f4109b3a26ad1cf51c7d66920ead67cb86bb7fc#diff-b1b3ce8fbe4a2f02397d6e5805ae0869fa7ada243daf0d58c97d7f73bccf0b7c).

finagolfin commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -273,12 +273,42 @@ add_custom_target(musl_modulemap DEPENDS ${musl_modulemap_target_list})
 set_property(TARGET musl_modulemap PROPERTY FOLDER "Miscellaneous")
 add_dependencies(sdk-overlay musl_modulemap)
 
+# Add a system 'math' module overlay, that only applies only to platforms that
```
that only applies only -> which only applies

hyp committed:
comment cleanup

hyp commented:
@swift-ci please test

hyp commented:
Planning to merge later today, as Ian's feedback regarding the overlays with math and float constants has been addressed, and I would like to merge it this week.

hyp commented:
follow-up Synchronization fix that just got broken: https://github.com/apple/swift/pull/74160
